src/lib/comms/include/sol-connman.h

```diff
@@ -0,0 +1,382 @@
+
+/**
+ * @enum sol_connman_state
+ *
+ * @brief the global connection state of system
+ *
+ * One of these must be chosen for the global connection state,
+ * with sol_connman_get_state()
+ *
+ */
+enum sol_connman_state {
+    SOL_CONNMAN_STATE_UNKNOWN   = 0,
+    SOL_CONNMAN_STATE_IDLE,
+    SOL_CONNMAN_STATE_READY,
+    SOL_CONNMAN_STATE_ONLINE,
+    SOL_CONNMAN_STATE_OFFLINE,
+};
+
+/**
+ * @brief Service monitor callback used to inform a service changed
+ *
+ * @param data the user data
+ * @param service the monitor service
+ */
+typedef void (*sol_connman_service_monitor_cb)
+    (void *data,
+    const struct sol_connman_service *service);
+
+/**
+ * @brief Manager monitor callback used to inform a manager updated
+ *
+ * @param data the user data
+ */
+typedef void (*sol_connman_manager_monitor_cb)(void *data);
+
+/**
+ * @brief error monitor callback used to inform a call result
+ *
+ * @param data the user data
+ * @param call_result the call result
+ */
+typedef void (*sol_connman_error_monitor_cb)(void *data, bool call_result);
+
+/**
+ * @brief Gets the service name
+ *
+ * This function gets the name for the connman service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the name is desired
+ *
+ * @return The name of the services on success, NULL on failure.
+ */
+const char *sol_connman_service_get_name(
+    const struct sol_connman_service *service);
+
+/**
+ * @brief Gets the service state
+ *
+ * This function gets the state for the connman service,
+ * after the service is to give by service monitor.
+ *
+ * @note SOL_CONNMAN_SERVICE_STATE_REMOVE is used to show
+ * the service has been removed. When it is notified,
+ * the service has been removed in the system.
+ *
+ * @param service The service structure which the state is desired
+ *
+ * @return The state of the services on success,
+ * SOL_CONNMAN_SERVICE_STATE_UNKNOWN on failure.
+ */
+enum sol_connman_service_state sol_connman_service_get_state(
+    const struct sol_connman_service *service);
+
+/**
+ * @brief Gets the service type
+ *
+ * This function gets the type for the connman service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the type is desired
+ *
+ * @return The type of the services on success,
+ * SOL_CONNMAN_SERVICE_TYPE_UNKNOWN on failure.
+ */
+const char *sol_connman_service_get_type(
+    const struct sol_connman_service *service);
+
+/**
+ * @brief Gets the service network address
+ *
+ * This function gets the network address for the connman service,
+ * after the service is to give by service monitor.
+ *
+ * @param service The service structure which the network address is desired
+ * @param link The retrieved content
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int sol_connman_service_get_network_address(
+    const struct sol_connman_service *service, struct sol_network_link **link);
```

Comment:
Please check the implementing of sol_connman_service_get_network_address.
SOL_API int
 sol_connman_service_get_network_address(const struct sol_connman_service *service,
     struct sol_network_link **link)
 {
     SOL_NULL_CHECK(service, -EINVAL);
     SOL_NULL_CHECK(link, -EINVAL);
     *link = (struct sol_network_link *)&service->link;
     return 0;
 }

I just want to transfer the pointer of  struct sol_network_link to the user.
Any idea? Thanks.